manager/src/main/scala/org/openinfralabs/caerus/cache/manager/BasicStorageIOMultiTierPlanner.scala

```diff
@@ -123,96 +124,82 @@ case class BasicStorageIOPlanner(optimizer: Optimizer, predictor: Predictor, pat
-        Some(CaerusDelete(deletedCandidates.toSeq.map(contents)))
-      } else {
-        None
+    var all_candidates: Seq[Candidate] = candidates
+    for(tier <- Tier.values){
+      if(all_contents.keySet.contains(tier)){
+        val contents = all_contents(tier)
+        logger.info("existing contents for Tier %s: %s\n".format(tier, contents.mkString("\n")))
+        logger.info("existing candidates size:\n%s\n".format(all_candidates.size))
+        val remainingCandidates: Seq[Candidate] = all_candidates.filter(!contents.contains(_)).filter(candidate => {
+          val tempReferences: mutable.HashMap[String,Long] = mutable.HashMap.empty[String,Long]
+          optimizer.optimize(plan, contents + ((candidate, "temp")), addReference(tempReferences))
+          tempReferences.contains("temp")
+        })
+        logger.info("Remaining candidates:\n%s".format(remainingCandidates.mkString("\n")))
+        if (remainingCandidates.nonEmpty) {
+          val initialCost: Long = findCost(contents, plans)
+          logger.info("Initial I/O: %s MiB".format(initialCost / (1024 * 1024)))
+
+          // Find candidate per byte benefit.
+          val costs: Seq[Long] = remainingCandidates.map(candidate => {
+            findCost(contents + ((candidate, "temp")), plans) +
+              findSourceReadSize(contents, candidate) +
+              candidate.sizeInfo.get.writeSize
+          })
+          val benefits: Seq[Long] = costs.map(cost => initialCost - cost)
+          val benefitsPerByte: Seq[Long] = benefits.indices.map(i => {
+            benefits(i) / remainingCandidates(i).sizeInfo.get.writeSize
+          })
+          logger.info("Benefits/Byte: %s".format(benefitsPerByte.mkString(", ")))
+
+          val topCandidateIndex: Int = benefitsPerByte.zipWithIndex.maxBy(_._1)._2
+          val topCandidate = remainingCandidates(topCandidateIndex)
+          var newCost: Long = costs(topCandidateIndex)
+          if (newCost - initialCost >= 0) {
+            logger.info("No positive candidate found. No content change")
+            new_multitier_contents(tier) = contents
+
+          } else {
+            logger.info("Top candidate: %s".format(topCandidate))
+
+            val topCandidateName: String = getName(topCandidate)
+            val topCandidatePath: String = tiers(tier) + Path.SEPARATOR + topCandidateName
+            var newContents = contents + ((topCandidate, topCandidatePath))
+            var newSize: Long = newContents.keys.map(candidate => candidate.sizeInfo.get.writeSize).sum
+            while (newSize > capacity(tier)) {
+              logger.warn("Capacity is not big enough to hold an extra candidate. Do not change anything.")
+              val bottomCandidateInfo: (Candidate, Long) = findBottomCandidate(newContents, plans, newCost)
+              val bottomCandidate: Candidate = bottomCandidateInfo._1
+              newCost = bottomCandidateInfo._2
+              newContents -= bottomCandidate
```

Comment:
A check is missing here. If the bottom candidate is the top candidate from before we should also return the old contents without change.

For example, let us assume we have capacity 3 and we have A with size 1 and B with size 1 in the cache.
If C with size 3 is selected as a new top candidate (now size is 5 and capacity is 3) this protocol might:
1) Remove A (size is 4 and capacity 3)
2) Remove C (size is 1 and capacity 1)

In this case, instead of keeping only B in the cache, since we do not actually insert C, we should keep both A and B in the cache (same to previous configuration).

Reply:
I see, I will only update contents after remove bottomcandidate if newContents.contains(topCandidate)